File: Python/ch02/MyArrayList.py
# 我的数组容器
class MyArrayList:
    def __init__(self, capacity):
        """

        :param capacity: 长度
        """
        self.array = [None] * capacity
        self.size = 0

    def add_index_element(self, index, element):
        """
        在指定位置添加元素
        :param index: 指定位置
        :param element: 元素
        :return:
        """
        # 判断访问下标是否超出范围
        if index < 0 or index > self.size:
            raise Exception("超出数组实际元素范围！")
        # 如果实际元素达到数组容量上线，数组扩容
        if self.size >= len(self.array):
            self.resize()
        # 从右向左循环，逐个元素向右挪一位。
        for i in range(self.size - 1, index - 1, -1):
            self.array[i + 1] = self.array[i]
        # 腾出的位置放入新元素
        self.array[index] = element
        self.size += 1

    def add_element(self, element):
        """
        在尾部放入新元素
        :param element: 放在尾部的元素
        :return:
        """
        # 如果实际元素达到数组容量上线，数组扩容
        if self.size >= len(self.array):
            self.resize()
        # 在尾部放入新元素
        self.array[self.size] = element
        self.size += 1

    def add(self, *args):
        """
        add
        :param args:
        :return:
        """
        if len(args) == 2:
            self.add_index_element(index=args[0], element=args[1])
        if len(args) == 1:
            self.add_element(element=args[0])

    def resize(self):
        """
        数组扩容
        :return:
        """
        array_new = [None] * len(self.array) * 2
        # 从旧数组拷贝到新数组
        for i in range(self.size):
            array_new[i] = self.array[i]
        self.array = array_new

    def remove(self, index):
        """
        移除指定位置的数组
        :param index: 指定的位置
        :return:
        """
        # 判断访问下标是否超出范围
        if index < 0 or index >= self.size:
            raise Exception("超出数组实际元素范围！")
        # 从左到右，逐个元素向左挪动一位
        for i in range(index, self.size):
            self.array[i] = self.array[i + 1]
        self.size -= 1

    def output(self):
        """
        输出数组
        :return:
        """
        print(self.array)


myArrayList = MyArrayList(5)
myArrayList.add(0, 1)
myArrayList.add(1, 2)
myArrayList.add(2, 3)
myArrayList.output()
myArrayList.add(4)
myArrayList.add(5)
myArrayList.add(6)
myArrayList.output()
myArrayList.remove(1)
myArrayList.remove(1)
myArrayList.remove(1)
myArrayList.output()

File: Java/src/ch02/MyArrayList.java
package ch02;

/**
 * 我的数组容器
 */
public class MyArrayList {

    // 用来存储容器中的元素
    private int[] array;
    // 容器中元素的个数
    private int size;

    /**
     * 构造方法
     * @param capacity 初始化大小
     */
    public MyArrayList(int capacity) {
        this.array = new int[capacity];
        size = 0;
    }

    /**
     * 在制定位置新增元素
     * @param index 制定的位置
     * @param element 元素
     * @throws Exception 数组越界异常
     */
    public void add(int index, int element) throws Exception {

        //判断访问下标是否超出范围
        if(index<0 || index>size){
            throw new IndexOutOfBoundsException("超出数组实际元素范围！");
        }

        //如果实际元素达到数组容量上线，数组扩容
        if(size >= array.length){
            resize();
        }

        //从右向左循环，逐个元素向右挪一位。
        for(int i=size-1; i>=index; i--){
            array[i+1] = array[i];
        }

        //腾出的位置放入新元素
        array[index] = element;

        size++;
    }

    /**
     * 在最近的一个空位添加元素
     * @param element 元素
     * @throws Exception
     */
    public void add(int element){
        //如果实际元素达到数组容量上线，数组扩容
        if(size >= array.length){
            resize();
        }

        //腾出的位置放入新元素
        array[size] = element;

        size++;
    }

    /**
     * 数组扩容
     */
    public void resize(){
        // 扩大到原来大小的两倍
        int[] arrayNew = new int[array.length*2];
        //从旧数组拷贝到新数组
        System.arraycopy(array, 0, arrayNew, 0, array.length);
        array = arrayNew;
    }

    /**
     * 在指定位置删除元素
     * @param index 制定位置
     * @return 被删除的元素
     * @throws Exception
     */
    public int remove(int index) throws Exception {
        //判断访问下标是否超出范围
        if(index<0 || index>=size){
            throw new IndexOutOfBoundsException("超出数组实际元素范围！");
        }

        int deletedElement = array[index];
        //从左向右循环，逐个元素向左挪一位。
        for(int i=index; i<(size-1); i++){
            array[i] = array[i+1];
        }
        size--;

        return deletedElement;
    }



    /**
     * 输出数组
     */
    public void output(){
        String rnt = "";
        for (int i = 0; i < size; i++) {
            rnt = rnt + array[i] + " ";
        }
        System.out.println(rnt);
    }

    public static void main(String[] args) throws Exception {
        MyArrayList myArrayList = new MyArrayList(3);
        myArrayList.add(0,1);
        myArrayList.add(1,2);
        myArrayList.add(2,3);
        myArrayList.output();
        myArrayList.add(4);
        myArrayList.add(5);
        myArrayList.add(6);
        myArrayList.output();
        myArrayList.remove(1);
        myArrayList.remove(1);
        myArrayList.remove(1);
        myArrayList.output();
    }
}